storage/aptosdb/src/transaction_store/mod.rs

```diff
@@ -150,6 +150,69 @@ impl TransactionStore {
         Err(AptosDbError::NotFound(format!("BlockMetadata preceding version {}", version)).into())
     }
 
+    /// Searches forward from the version to find the state checkpoint transaction
+    pub fn get_block_boundaries(&self, version: Version) -> Result<(Version, Version)> {
+        // Must be larger than a block size, otherwise a NotFound error will be raised wrongly.
+        const MAX_VERSIONS_TO_SEARCH: usize = 1000 * 100;
+
+        // Genesis is always 0,0
+        if version == 0 {
+            return Ok((0, 0));
+        }
+
+        // Linear searches via `DB::iter()` and `DB::rev_iter()` here, NOT expecting performance hit, due to the fact
+        // that the iterator caches data block and that there are limited number of transactions in
+        // each block.
+        let mut iter = self.db.rev_iter::<TransactionSchema>(Default::default())?;
+        iter.seek(&version)?;
+        let mut start_version = None;
+        for res in iter.take(MAX_VERSIONS_TO_SEARCH) {
+            let (v, txn) = res?;
+            // If we've found the beginning of the block, stop
+            if matches!(
+                txn,
+                Transaction::GenesisTransaction(_) | Transaction::BlockMetadata(_)
+            ) {
+                start_version = Some(v);
+                break;
+            }
+        }
```

Comment:
Expecting genesis txn after genesis is kinda messy.. 

Shall we emit NewBlockEvent at both types of transactions and find block boundaries with that? @zekun000 
Also, can we assume the sequence number on the NewBlockEvent to be the block height? Or put the height in the event? @zekun000 
if the seq_num is the block height, we can get_block_by_block_index() using this index directly https://github.com/aptos-labs/aptos-core/blob/b1601d23fb157fee4c4932cd093d3809be8094a4/storage/aptosdb/src/schema/event_by_key/mod.rs#L10

we have this index https://github.com/aptos-labs/aptos-core/blob/b1601d23fb157fee4c4932cd093d3809be8094a4/storage/aptosdb/src/schema/event_by_version/mod.rs#L10 to find the sequence number for a specific event around a certain version directly, without a linear search. (get_block_metadata() can be updated to use it.)